Umbraco 15.x-16.x-17.0.0-rc2 .NET 9.0 / .NET 10.0 support

gavinfaux opened the pull request:
* Bump Node requirement to 24+
* Client package tool changed to pnpm 10.20+ for enhanced CI workflow, e.g. prevent installed dependencies from running scripts
* .NET 9.0 / 10.0 support, add support for Umbraco 17.0.0-rc2
* Use Umbraco supported @hey-api/openapi-ts
* Adds back office cookie authentication service , extracted from Material theme
* Extends Management API with public OpenIddict client
* Front end Markdown Editor /a-new/ route, OAuth 2.0 Authorization Code flow with PKCE
* API Refactors to remove complexity, support updated targets, enhance security and compatability
* Refactor build pipeline - cross-platform support (WSL/Linux/macOS), GitHub linux build action
* Articulate.StaticAssets project: output of client back office extension and other static RCL assets, gated behind opt in `EnableClientBuild` variable (disabled by default to prevent Visual Studio incremental build issues), enabled for CI builds

Removed:
  * Razor Compilation at runtime for RCL view reload dx - no official support going forward with .NET 10, dotnet hot reload intented to replace
  * Removed unnecessary tracking of Umbraco settings JSON schema
  * Remove incomplete ASP.NET Core based theming additions (move to a future PR)

gavinfaux commented:
Moved to draft mode while:
- [x]  pending review/rebase for upstream changes
- [x]  review responsive Markdown Editor features, happy path verified, test error scenarios with client and API
- [x]  review PKCE logout auth flow to ensure full backoffice logout
- [x]  review backwards compatability with legacy data/content formats (media, blog export / import feature)

Shazwazza commented:
@gavinfaux do you think its going to be realistic to support both umb 16 + 17 with the single build? Or should we just go straight to Umb 17 LTS?

gavinfaux commented:
> @gavinfaux do you think its going to be realistic to support both umb 16 + 17 with the single build? Or should we just go straight to Umb 17 LTS?

ok so that was before [v17 should use swashbuckle 10](https://github.com/umbraco/Umbraco-CMS/discussions/20902), but now resolved

gavinfaux commented:
- [x] v17.0 release compatible, + v15/v16; I know v15 is eol now but it works as is, can remove if required
- [x] linux / wsl build and workflow updates, windows-latest workflow paused, ubuntu-latest workflow enabled
- [x] Markdown Editor available @ /a-new
- [x] switched to pnpm vs npm to disallow post install scripts security incidents; updated .npmrc `ignore-scripts` for npm

- [x] Review CSP headers for markdown editor and default themes, move to docs?

- [ ] Review backwards compatibility v17 auth related updates since `17.0.0-rc2` to [17.0.1](https://releases.umbraco.com/release/umbraco/Umbraco-CMS/17.0.1)
  - [x] v15/v16: feature flag defaults useCookieAuth=false, so bearer mode continues to work.
  - [ ] v17+: useCookieAuth=true, bearer dropped, cookies handle auth.
  - [ ] if useCookieAuth, skip storing/reading tokens, drop Authorization headers, keep credentials: 'include' on fetch, simplify logout; else keep the existing bearer-token flow.

- [x] ⏸️ hold: ImageCropperValue property value migration from v1 to v3 + `/media/path/file.ext` to v3 or amend export
